File: packages/python/src/agent_capture/redaction/patterns_finance.py
"""Finance-specific pattern recognizers.

Each recognizer detects one kind of sensitive value in free text. The
recognizer returns the list of matches it found; the filter pipeline
decides how to replace them per the customer's policy.

Coverage (architecture doc §8.1):

- US SSN — ``123-45-6789`` and ``123456789``
- ABA routing number — 9 digits passing the ABA mod-10 checksum
- US bank account — heuristic for 8-17 digit sequences in account context
- MICR line — magnetic ink character recognition line on a US check
- Date of birth — common written forms (best-effort)

Recognizers are intentionally conservative — false positives over-redact
(safe), false negatives miss real PII (unsafe). When in doubt the
recognizer fires.
"""

from __future__ import annotations

import re
from collections.abc import Callable
from dataclasses import dataclass
from typing import Final


@dataclass(frozen=True)
class Match:
    """A single hit from a recognizer."""

    start: int
    end: int
    value: str
    recognizer: str


@dataclass(frozen=True)
class Recognizer:
    """A named regex (or callable) detector with a field-type label.

    ``field_type`` is the label embedded in the redaction output (e.g.
    ``[REDACTED:ssn]``) and the key the policy looks up to find which
    strategy to apply.
    """

    name: str
    field_type: str
    pattern: re.Pattern[str]
    validator: Callable[[str], bool] | None = None

    def find_all(self, text: str) -> list[Match]:
        out: list[Match] = []
        for m in self.pattern.finditer(text):
            captured = m.group(0)
            if self.validator is not None and not self.validator(captured):
                continue
            out.append(Match(start=m.start(), end=m.end(), value=captured, recognizer=self.name))
        return out


# ---- ABA routing number checksum -----------------------------------------


def _aba_checksum_valid(s: str) -> bool:
    """ABA routing number mod-10 check.

    Sum 3*(d0+d3+d6) + 7*(d1+d4+d7) + 1*(d2+d5+d8) must be divisible
    by 10. See: https://en.wikipedia.org/wiki/ABA_routing_transit_number
    """
    digits = [int(c) for c in s if c.isdigit()]
    if len(digits) != 9:
        return False
    weights = [3, 7, 1, 3, 7, 1, 3, 7, 1]
    return sum(d * w for d, w in zip(digits, weights, strict=True)) % 10 == 0


# ---- Recognizer pack ----------------------------------------------------


US_SSN: Final[Recognizer] = Recognizer(
    name="us_ssn",
    field_type="ssn",
    # Standard formats. Excludes obviously-invalid groups (000, 666, 9xx area;
    # 00 group; 0000 serial) per SSA rules to reduce false positives.
    pattern=re.compile(
        r"\b(?!000|666|9\d{2})\d{3}-(?!00)\d{2}-(?!0000)\d{4}\b"
        r"|"
        r"\b(?!000|666|9\d{2})\d{3}(?!00)\d{2}(?!0000)\d{4}\b"
    ),
)


ABA_ROUTING: Final[Recognizer] = Recognizer(
    name="aba_routing",
    field_type="routing_number",
    pattern=re.compile(r"\b\d{9}\b"),
    validator=_aba_checksum_valid,
)


US_BANK_ACCOUNT: Final[Recognizer] = Recognizer(
    name="us_bank_account",
    field_type="account_number",
    # 8-17 digits, optionally hyphenated. Conservative: only fires when
    # surrounded by acct/account context to limit false positives.
    pattern=re.compile(
        r"(?i)\b(?:account(?:\s+(?:number|no\.?|#))?|acct(?:\.\s*(?:no\.?|#))?)\s*[:#]?\s*"
        r"(?P<acct>\d[\d-]{7,16}\d)\b"
    ),
)


MICR_LINE: Final[Recognizer] = Recognizer(
    name="micr_line",
    field_type="micr",
    # MICR symbols ⑆ (transit), ⑈ (on-us), ⑉ (amount), ⑇ (dash). When
    # encoded as ASCII (Magtek devices use 'A'/'B'/'C'/'D'), keep both.
    pattern=re.compile(
        r"[⑆⑇⑈⑉][^⑆⑇⑈⑉]+[⑆⑇⑈⑉]"
        r"|"
        r"A\d{9}A\s*[A-D\d-]{5,}",
    ),
)


DOB: Final[Recognizer] = Recognizer(
    name="dob",
    field_type="date_of_birth",
    # Common written DOB forms. Fires only with a "DOB:"/"born"/"D.O.B." cue
    # to avoid redacting every date in a document.
    pattern=re.compile(
        r"(?i)\b(?:dob|d\.o\.b\.?|date\s+of\s+birth|born(?:\s+on)?)\b"
        r"[:\s-]*"
        r"(?P<date>"
        r"(?:0?[1-9]|1[0-2])[/-](?:0?[1-9]|[12]\d|3[01])[/-](?:19|20)\d{2}"
        r"|"
        r"(?:19|20)\d{2}[/-](?:0?[1-9]|1[0-2])[/-](?:0?[1-9]|[12]\d|3[01])"
        r")"
    ),
)


DEFAULT_RECOGNIZERS: Final[tuple[Recognizer, ...]] = (
    US_SSN,
    ABA_ROUTING,
    US_BANK_ACCOUNT,
    MICR_LINE,
    DOB,
)

File: packages/typescript/src/redaction/patternsFinance.ts
/**
 * Finance-specific pattern recognizers. Mirrors
 * agent_capture.redaction.patterns_finance.
 *
 * Coverage:
 * - US SSN — dashed and undashed, SSA area-rule exclusions
 * - ABA routing — 9 digits, mod-10 checksum validated
 * - US bank account — context-gated 8-17 digits
 * - MICR line — basic ASCII encoding
 * - DOB — cue-word-gated dates
 */

export interface Match {
  start: number;
  end: number;
  value: string;
  recognizer: string;
}

export interface Recognizer {
  name: string;
  fieldType: string;
  findAll(text: string): Match[];
}

function abaChecksumValid(s: string): boolean {
  const digits = [...s].filter((c) => /\d/.test(c)).map((c) => parseInt(c, 10));
  if (digits.length !== 9) return false;
  const weights = [3, 7, 1, 3, 7, 1, 3, 7, 1];
  let sum = 0;
  for (let i = 0; i < 9; i++) sum += digits[i]! * weights[i]!;
  return sum % 10 === 0;
}

function regexRecognizer(
  name: string,
  fieldType: string,
  pattern: RegExp,
  validator?: (matched: string) => boolean,
): Recognizer {
  return {
    name,
    fieldType,
    findAll(text: string): Match[] {
      const out: Match[] = [];
      const re = new RegExp(pattern.source, pattern.flags.includes("g") ? pattern.flags : pattern.flags + "g");
      let m: RegExpExecArray | null;
      while ((m = re.exec(text)) !== null) {
        if (validator !== undefined && !validator(m[0])) continue;
        out.push({ start: m.index, end: m.index + m[0].length, value: m[0], recognizer: name });
        if (m[0].length === 0) re.lastIndex++; // guard against zero-width loops
      }
      return out;
    },
  };
}

export const US_SSN: Recognizer = regexRecognizer(
  "us_ssn",
  "ssn",
  // Standard formats, excluding obviously-invalid groups (000, 666, 9xx area; 00 group; 0000 serial).
  /\b(?!000|666|9\d{2})\d{3}-(?!00)\d{2}-(?!0000)\d{4}\b|\b(?!000|666|9\d{2})\d{3}(?!00)\d{2}(?!0000)\d{4}\b/g,
);

export const ABA_ROUTING: Recognizer = regexRecognizer(
  "aba_routing",
  "routing_number",
  /\b\d{9}\b/g,
  abaChecksumValid,
);

export const US_BANK_ACCOUNT: Recognizer = regexRecognizer(
  "us_bank_account",
  "account_number",
  /\b(?:account(?:\s+(?:number|no\.?|#))?|acct(?:\.\s*(?:no\.?|#))?)\s*[:#]?\s*(\d[\d-]{7,16}\d)\b/gi,
);

export const MICR_LINE: Recognizer = regexRecognizer(
  "micr_line",
  "micr",
  /[⑆⑇⑈⑉][^⑆⑇⑈⑉]+[⑆⑇⑈⑉]|A\d{9}A\s*[A-D\d-]{5,}/g,
);

export const DOB: Recognizer = regexRecognizer(
  "dob",
  "date_of_birth",
  /\b(?:dob|d\.o\.b\.?|date\s+of\s+birth|born(?:\s+on)?)\b[:\s-]*((?:0?[1-9]|1[0-2])[/-](?:0?[1-9]|[12]\d|3[01])[/-](?:19|20)\d{2}|(?:19|20)\d{2}[/-](?:0?[1-9]|1[0-2])[/-](?:0?[1-9]|[12]\d|3[01]))/gi,
);

export const DEFAULT_RECOGNIZERS: ReadonlyArray<Recognizer> = [
  US_SSN,
  ABA_ROUTING,
  US_BANK_ACCOUNT,
  MICR_LINE,
  DOB,
];
